pep-9999.rst

```diff
@@ -0,0 +1,107 @@
+deprecating and then disabling new uploads of distribution filenames that have
+the ``.egg`` suffix.
+
+It does not recommend removing or otherwise affecting any previously
+uploaded ``bdist_egg`` distributions.
+
+Rationale
+=========
+
+Previous Work
+-------------
+
+The groundwork for this proposal was established with :pep:`527`, which
+proposed deprecating and eventually removing upload support for a handful
+of un(der)used file extensions and distribution types.
+
+In particular, :pep:`527` proposed the removal of the ``bdist_dumb``,
+``bdist_rpm``, ``bdist_dmg``, ``bdist_msi``, and ``bdist_wininst`` distribution
+types, as well as the ``.tar``, ``.tar.bz2``, ``.tar.xz``, ``.tar.Z``,
+``.tgz``, and ``.tbz`` file extensions on distribution filenames.
+
+:pep:`527` was fully enacted with
+`PR #7529 <https://github.com/pypi/warehouse/pull/7529>`_ to Warehouse,
+which was merged on 13-Apr-2020.
+
+The ``bdist_egg`` format
+------------------------
+
+The ``bdist_egg`` filetype identifies distributions in the
+`Egg format <https://packaging.python.org/en/latest/glossary/#term-Egg>`_. The
+Egg format was introduced by setuptools in 2004 and is roughly equivalent
+in functionality to the
+`Wheel format <https://packaging.python.org/en/latest/glossary/#term-Wheel>`_
+introduced and standardized by :pep:`427` in 2012.
+
+Despite their longevity, the Egg format has had
+`limited adoption on PyPI <https://github.com/pypi/warehouse/issues/10653>`_.
+Some observations from that issue:
+
+* As of December 2021, ``bdist_egg`` uploads account for less than 1% of all
+  built distribution uploads;
+* ``pip`` deprecated its ``--egg`` option in 2016;
+* ``setuptools`` has considered Egg support deprecated since 2019;
+* ``build`` only supports the ``sdist`` and ``bdist_wheel`` filetypes.
+
+Given the above, this PEP proposes the removal of the ``bdist_egg`` format
+under the same justifications presented in :pep:`527`, namely:
+
+1. Egg distributions are of of limited use to the broader ecosystem and
+   therefore represent a non-reciprocal maintenance burden;
+2. Continued support for Egg distributions represents a bandwidth and storage
+   cost to PyPI, as well as to any external mirrors of PyPI.
+
+This PEP *additionally* offers an argument for removal rooted in
+standardization and duplication: the Egg format is not standardized by any
+PEP or other community standard, and overlaps heavily with its standardized
+and well-supported alternative (Wheel).
+
+The ``.egg`` file extension
+---------------------------
+
+The ``.egg`` file extension is used exclusively for distributions of the
+``bdist_egg`` format. As such, it serves no purpose in a scenario where
+PyPI disables new distribution uploads for ``bdist_egg`` distributions.
+
+Removal Process
+===============
+
+This PEP proposes a removal process based on :pep:`527`, but with the following
+modifications:
+
+* The deprecation period will be six months, rather than one month.
```

Comment:
Suggestion: Stick with one month. There was outreach done in the initial warehouse issue to determine outstanding usecases of egg uploads. At that time eggs were already on the decline for uploads.

Question: Would updated metrics/stats for egg uploads the helpful to motivate moving this to a more accelerated timeline?

Reply:
Sounds good to me -- I arbitrarily chose 6 months because of the more "aggressive" email policy (don't email all current egg uploaders, only upload if they upload another egg), but IMO 1 month is also suitable 🙂

Reply:
fb4c3847

Reply:
Volumes are low enough I'm actually game to do the one-time email with a one month deprecation.

Reply:
Peak .egg was `2020-05` with `1534` uploaded files.

Here's the last 12 full months for .egg files uploaded:

```
 2022-12 |   531
 2023-01 |   653
 2023-02 |   695
 2023-03 |   570
 2023-04 |   620
 2023-05 |   576
```

Reply:
all 1022 projects and the number of eggs they've uploaded, in 2022: https://gist.github.com/ewdurbin/d7939458f6205069adeeb18c454c9331

Reply:
I'm in favor of a one-time bulk email + a 1 mo deprecation period with emails per upload.

Reply:
92a85aa3